File: backends/java/zappy-application/src/main/java/com/zappymart/application/ports/ProductRepository.java
package com.zappymart.application.ports;

import com.zappymart.domain.catalogue.Product;

import java.util.Collection;
import java.util.List;
import java.util.Optional;

public interface ProductRepository {

    List<Product> catalogue();

    Optional<Product> byId(String productId);

    Optional<Product> bySlug(String slug);

    List<Product> byIds(Collection<String> productIds);

    void reduceStock(String productId, int quantity);
}

File: backends/kotlin/zappy-application/src/main/kotlin/nl/zappymart/application/ports/ProductRepository.kt
package nl.zappymart.application.ports

import nl.zappymart.domain.catalogue.Product
import nl.zappymart.domain.catalogue.ProductSpecification

interface ProductRepository {

    fun page(specification: ProductSpecification, size: Int, afterProductId: String?): List<Product>

    fun count(specification: ProductSpecification): Int

    fun findById(productId: String): Product?

    fun findBySlug(slug: String): Product?

    fun findAllByIds(productIds: List<String>): List<Product>

    fun reduceStock(quantityPerProductId: Map<String, Int>)
}
